### PR TITLE
variable solver data concept

### PR DESCRIPTION
**Pull request by Affie**
draft to help with the discussion in #182 concept

### CODE REVIEW

**Comment by dehann** [comment]
will look and answer through review, thanks for getting this going!

File: src/entities/DFGVariable.jl [hunk 1]
```diff
@@ -55,3 +54,129 @@ DFGVariable(label::Symbol, softtype::T;
...
+
+"""
+    $(SIGNATURES)
+Structure for first-class citizens of a DFGVariable.
+"""
+mutable struct DFGVariableSummary <: AbstractDFGVariable
+    label::Symbol
+    timestamp::DateTime
+    tags::Vector{Symbol}
+    estimateDict::Dict{Symbol, <:AbstractPointParametricEst}
```
Comment (dehann) [inline src/entities/DFGVariable.jl]: latest v0.5.3 has this as `ppeDict`

Comment (dehann) [reply src/entities/DFGVariable.jl]: see here: https://github.com/JuliaRobotics/DistributedFactorGraphs.jl/blob/bea526695ad7f55ab695fd80b34dd74e3bcd60d6/src/entities/DFGVariable.jl#L186

Comment (dehann) [reply src/entities/DFGVariable.jl]: will likely fall out naturally when merging all this to master, just ask to keep an eye on it when we do please.

File: src/entities/DFGVariable.jl [hunk 2]
```diff
@@ -55,3 +54,129 @@ DFGVariable(label::Symbol, softtype::T;
...
+const VariableDataLevel1 = Union{DFGVariable, DFGVariableSummary}
+const VariableDataLevel2 = Union{DFGVariable}
+
+
+"""
+$SIGNATURES
+
+Return the estimates for a variable.
+"""
+getEstimates(v::VariableDataLevel1) = v.estimateDict
```
Comment (dehann) [inline src/entities/DFGVariable.jl]: This too has been deprecated, see master branch.

File: src/needsahome.jl [hunk 3]
```diff
@@ -59,59 +72,61 @@ function buildSubgraphFromLabels(dfg::G,
...
+#
+# DevNotes
+# - DF: Could we somehow better consolidate the functionality of this method into `buildSubgraphFromLabels!` above, which in turn should be consolidated as SamC suggests.
+# - Since this function has happened more than once, it seems the name `buildSubgraphFromLabels!` might stick around, even if it just becomes a wrapper.
+#
+# Related
+#
+# buildSubgraphFromLabels!, _copyIntoGraph!, getVariableIds
+# """
+# function buildSubgraphFromLabels!_SPECIAL(dfg::G,
```
Comment (dehann) [inline src/needsahome.jl]: Just and FYI:  @Affie, this is your add hoc function for subGraph from a while back -- see #250

File: src/services/AbstractDFG.jl [hunk 4]
```diff
@@ -432,6 +432,82 @@ function _copyIntoGraph!(sourceDFG::G, destDFG::H, variableFactorLabels::Vector{
     return nothing
 end
 
+#TODO UNTESTED and NOT FILLED IN, just to define function signatures
+"""
+    $(SIGNATURES)
+"""
+function getVariableSolverData(dfg::AbstractDFG, variablekey::Symbol, solvekey::Symbol=:default)
+    error("not implemented")
+    return deepcopy(solverData(getVariable(dfg, variablekey), solvekey))
```
Comment (dehann) [inline src/services/AbstractDFG.jl]: why a deepcopy?

File: src/services/AbstractDFG.jl [hunk 5]
```diff
@@ -432,6 +432,82 @@ function _copyIntoGraph!(sourceDFG::G, destDFG::H, variableFactorLabels::Vector{
...
+
+    var = getVariable(dfg, variablekey)
+
+    if haskey(var.solverDataDict, solvekey)
+        error("VariableNodeData '$(solvekey)' already exists")
+    end
+
+    #var.solverDataDict[solvekey] = vnd
+
+    setSolverData(var, deepcopy(vnd), key)
```
Comment (dehann) [inline src/services/AbstractDFG.jl]: see updated api in master to include an `!`

File: src/services/AbstractDFG.jl [hunk 6]
```diff
@@ -432,6 +432,82 @@ function _copyIntoGraph!(sourceDFG::G, destDFG::H, variableFactorLabels::Vector{
...
+    $(SIGNATURES)
+"""
+function updateVariableSolverData!(dfg::AbstractDFG, vnd::VariableNodeData, variablekey::Symbol, solvekey::Symbol=:default)
+    error("not implemented")
+    #This is basically just setSolverData with a copy
+    var = getVariable(dfg, variablekey)
+
+    #var.solverDataDict[solvekey] = deepcopy(vnd)
+    #for InMemoryDFGTypes, cloud would update here
+    setSolverData(var, deepcopy(vnd), key)
```
Comment (dehann) [inline src/services/AbstractDFG.jl]: Also, why do a deep copy here?  You think the user might accidentally point the code to the same vnd object and therefore become incorrect?

File: src/services/AbstractDFG.jl [hunk 7]
```diff
@@ -432,6 +432,82 @@ function _copyIntoGraph!(sourceDFG::G, destDFG::H, variableFactorLabels::Vector{
...
+function deleteVariableSolverData!(dfg::AbstractDFG, variablekey::Symbol, solvekey::Symbol=:default)
+    error("not implemented")
+
+    var = getVariable(dfg, variablekey)
+
+    if haskey(var.solverDataDict, solvekey)
+        error("VariableNodeData '$(solvekey)' already exists")
+    end
+
+    vnd = pop!(var.solverDataDict, solvekey)
```
Comment (dehann) [inline src/services/AbstractDFG.jl]: This one seems right to me, will think on it more.

File: src/services/AbstractDFG.jl (hunk 4 above)
Comment (dehann) [reply src/services/AbstractDFG.jl]: should verb `get` not return the original memory object?

**Comment by Affie** [comment]
Closing because of worry about merges. I will cherry pick or manually copy the relevant code.

File: src/services/AbstractDFG.jl (hunk 6 above)
Comment (Affie) [reply src/services/AbstractDFG.jl]: I think we should maybe leave the ‘deepcopy’ responsibility to the user.

File: src/services/AbstractDFG.jl (hunk 4 above)
Comment (Affie) [reply src/services/AbstractDFG.jl]: Yes, I think you are correct.

**Pull request closed without merging**